File: scripts/pack_orb.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

ROOT = Dir.pwd
SRC_DIR = File.join(ROOT, 'src')

def read_trimmed(file_path)
  File.read(file_path).sub(/\s+\z/, '')
end

def indent(content, spaces)
  content.lines(chomp: true).map { |line| line.empty? ? '' : "#{' ' * spaces}#{line}" }.join("\n")
end

def expand_includes(content, base_dir = SRC_DIR)
  expanded = content.gsub(/^(\s*[^:\n]+:\s*)<<include\(([^)]+)\)>>\s*$/) do
    prefix = Regexp.last_match(1)
    include_path = Regexp.last_match(2).strip
    included = read_trimmed(File.join(base_dir, include_path))
    key_indent = prefix[/\A\s*/].length
    "#{prefix}|\n#{indent(included, key_indent + 2)}"
  end

  expanded.gsub(/^(\s*)<<include\(([^)]+)\)>>/) do
    leading = Regexp.last_match(1)
    include_path = Regexp.last_match(2).strip
    indent(read_trimmed(File.join(base_dir, include_path)), leading.length)
  end
end

def append_section(lines, section_name, directory_name)
  directory = File.join(SRC_DIR, directory_name)
  return unless Dir.exist?(directory)

  files = Dir.children(directory).select { |file| file.end_with?('.yml') }.sort
  return if files.empty?

  lines << ''
  lines << "#{section_name}:"
  files.each do |file|
    key = File.basename(file, '.yml')
    content = expand_includes(read_trimmed(File.join(directory, file)))
    lines << "  #{key}:"
    lines << indent(content, 4)
  end
end

lines = [expand_includes(read_trimmed(File.join(SRC_DIR, '@orb.yml')))]
append_section(lines, 'executors', 'executors')
append_section(lines, 'commands', 'commands')
append_section(lines, 'jobs', 'jobs')
append_section(lines, 'examples', 'examples')

File.write(File.join(ROOT, 'orb.yml'), "#{lines.join("\n")}\n")

File: scripts/pack_orb.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';

const root = process.cwd();
const srcDir = path.join(root, 'src');

const read = (filePath) => fs.readFileSync(filePath, 'utf8').replace(/\s+$/, '');

const indent = (content, spaces) =>
  content
    .split('\n')
    .map((line) => (line ? `${' '.repeat(spaces)}${line}` : ''))
    .join('\n');

const expandIncludes = (content, baseDir = srcDir) =>
  content
    .replace(/^(\s*[^:\n]+:\s*)<<include\(([^)]+)\)>>\s*$/gm, (_match, prefix, includePath) => {
      const included = read(path.join(baseDir, includePath.trim()));
      const keyIndent = prefix.match(/^\s*/)[0].length;
      return `${prefix}|\n${indent(included, keyIndent + 2)}`;
    })
    .replace(/^(\s*)<<include\(([^)]+)\)>>/gm, (_match, leading, includePath) => {
      const included = read(path.join(baseDir, includePath.trim()));
      return indent(included, leading.length);
    });

const appendSection = (lines, sectionName, directoryName) => {
  const directory = path.join(srcDir, directoryName);
  if (!fs.existsSync(directory)) {
    return;
  }

  const files = fs.readdirSync(directory).filter((file) => file.endsWith('.yml')).sort();
  if (files.length === 0) {
    return;
  }

  lines.push('', `${sectionName}:`);
  for (const file of files) {
    const key = path.basename(file, '.yml');
    const content = expandIncludes(read(path.join(directory, file)));
    lines.push(`  ${key}:`, indent(content, 4));
  }
};

const lines = [expandIncludes(read(path.join(srcDir, '@orb.yml')))];
appendSection(lines, 'executors', 'executors');
appendSection(lines, 'commands', 'commands');
appendSection(lines, 'jobs', 'jobs');
appendSection(lines, 'examples', 'examples');

fs.writeFileSync(path.join(root, 'orb.yml'), `${lines.join('\n')}\n`);
